maskit/masks.py

```diff
@@ -0,0 +1,296 @@
+    """
+
+    __slots__ = "_mask"
+
+    def __setitem__(self, key, value: bool):
+        """
+        Convenience function to set the value of a specific position of the
+        encapsulated :py:attr:`~.mask`.
+        """
+        if isinstance(key, int) or isinstance(key, slice) or isinstance(key, tuple):
+            self._mask[key] = value
+        else:
+            raise NotImplementedError
+
+    def __getitem__(self, key):
+        """
+        Convenience function to get the value of a specific position of the
+        encapsulated :py:attr:`~.mask`.
+        """
+        if isinstance(key, int) or isinstance(key, slice) or isinstance(key, tuple):
+            return self._mask[key]
+        raise NotImplementedError
+
+    @property
+    def mask(self):
+        """
+        Returns the encapsulated :py:attr:`~.mask`
+        """
+        return self._mask
+
+    def apply_mask(self, values: np.ndarray):
+        """
+        Applies the encapsulated py:attr:`~.mask` to the given ``values``.
+        Note that the values should have the same shape as the py:attr:`~.mask`.
+
+        :param values: Values where the mask should be applied to
+        """
+        raise NotImplementedError
+
+    def reset(self) -> None:
+        """Resets the mask to not mask anything."""
+        self._mask = np.zeros_like(self._mask, dtype=bool, requires_grad=False)
+
+    def perturb(
+        self,
+        amount: Optional[int] = None,
+        mode: PerturbationMode = PerturbationMode.INVERT,
+    ):
+        """
+        Perturbs the MaskedObject by the given ``mode`` of type
+        :py:class:`~.PerturbationMode` ``amount`` times. If no amount is given,
+        that is ``amount=None``, a random ``amount`` is determined given by the
+        actual size of the py:attr:`~.mask`. The ``amount`` is automatically
+        limited to the actual size of the py:attr:`~.mask`.
+
+        :param amount: Number of items to perturb, defaults to None
+        :param mode: How to perturb, defaults to PerturbationMode.INVERT
+        :raises AssertionError: Raised for negative amounts
+        :raises NotImplementedError: Raised in case of an unknown mode
+        """
+        assert (
+            amount is None or amount >= 0
+        ), "Negative values are not supported, please use PerturbationMode.REMOVE"
+        if amount == 0:
+            return
+        count = (
+            abs(amount) if amount is not None else rand.randrange(0, self._mask.size)
+        )
+        if mode == PerturbationMode.ADD:
+            indices = np.argwhere(~self._mask)
+        elif mode == PerturbationMode.INVERT:
+            indices = np.array([list(index) for index in np.ndindex(*self._mask.shape)])
+        elif mode == PerturbationMode.REMOVE:
+            indices = np.argwhere(self._mask)
+        else:
+            raise NotImplementedError(f"The perturbation mode {mode} is not supported")
+        if len(indices) == 0:
```

Comment:
Not sure if this is an issue, but consider that ``len`` for empty multi-dimensional arrays is *not* 0. Compare ``len(np.array([[]]))``. See [How can I check whether a numpy array is empty or not?](https://stackoverflow.com/q/11295609/5349916).

Reply:
Pingeridoo! 🦜

Reply:
Done!